account-compression/programs/account-compression/src/lib.rs

```diff
@@ -173,69 +177,149 @@ pub mod spl_account_compression {
+    /// canopy with the necessary nodes. The expected flow for a tree without canopy is
+    /// `prepare_tree` -> `init_merkle_tree_with_root`. The expected flow for a tree with canopy
+    /// is `prepare_tree` -> `append_canopy_nodes` (multiple times until all of the canopy is
+    /// filled) -> `init_merkle_tree_with_root`. This instruction initializes the tree header
+    /// while leaving the tree itself uninitialized. This allows distinguishing between an empty
+    /// tree and a tree prepare to be initialized with a root.
+    pub fn prepare_tree(
+        ctx: Context<Initialize>,
+        max_depth: u32,
+        max_buffer_size: u32,
+    ) -> Result<()> {
+        require_eq!(
+            *ctx.accounts.merkle_tree.owner,
+            crate::id(),
+            AccountCompressionError::IncorrectAccountOwner
+        );
+        let mut merkle_tree_bytes = ctx.accounts.merkle_tree.try_borrow_mut_data()?;
+
+        let (mut header_bytes, rest) =
+            merkle_tree_bytes.split_at_mut(CONCURRENT_MERKLE_TREE_HEADER_SIZE_V1);
+
+        let mut header = ConcurrentMerkleTreeHeader::try_from_slice(header_bytes)?;
+        header.initialize(
+            max_depth,
+            max_buffer_size,
+            &ctx.accounts.authority.key(),
+            Clock::get()?.slot,
+        );
+        header.serialize(&mut header_bytes)?;
+        let merkle_tree_size = merkle_tree_get_size(&header)?;
+        let (_tree_bytes, canopy_bytes) = rest.split_at_mut(merkle_tree_size);
+        check_canopy_bytes(canopy_bytes)
+    }
+
+    /// This instruction pre-initializes the canopy with the specified leaf nodes of the canopy.
+    /// This is intended to be used after `prepare_tree` and in conjunction with the
+    /// `init_merkle_tree_with_root` instruction that'll finalize the tree initialization.
+    pub fn append_canopy_nodes(
+        ctx: Context<Modify>,
+        start_index: u32,
+        canopy_nodes: Vec<[u8; 32]>,
+    ) -> Result<()> {
+        require_eq!(
+            *ctx.accounts.merkle_tree.owner,
+            crate::id(),
+            AccountCompressionError::IncorrectAccountOwner
+        );
+        let mut merkle_tree_bytes = ctx.accounts.merkle_tree.try_borrow_mut_data()?;
+
+        let (header_bytes, rest) =
+            merkle_tree_bytes.split_at_mut(CONCURRENT_MERKLE_TREE_HEADER_SIZE_V1);
+
+        let header = ConcurrentMerkleTreeHeader::try_from_slice(header_bytes)?;
+        header.assert_valid_authority(&ctx.accounts.authority.key())?;
+        // assert the tree is not initialized yet, we don't want to overwrite the canopy of an
+        // initialized tree
+        let merkle_tree_size = merkle_tree_get_size(&header)?;
+        let (tree_bytes, canopy_bytes) = rest.split_at_mut(merkle_tree_size);
+        // ensure the tree is not initialized, the hacky way
+        require!(
+            tree_bytes.iter().all(|&x| x == 0),
+            AccountCompressionError::TreeAlreadyInitialized
+        );
```

Comment:
Should something like this also be in `prepare_tree()`?  In the `Initialize` account validation struct used for `prepare_tree()`, it indicates the account will be zeroed out but I don't actually see where that is done.  If it is a gap, its probably an existing gap with `init_empty_merkle_tree()` as well.

Reply:
Ok we see the zero anchor validation, not an issue